plugin/pkg/admission/noderestriction/admission.go

```diff
@@ -487,26 +481,8 @@ func getLabelNamespace(key string) string {
 	return ""
 }
 
-// getForbiddenCreateLabels returns the set of labels that may not be set by the node.
-// TODO(liggitt): in 1.19, expand to match getForbiddenUpdateLabels()
-func (p *Plugin) getForbiddenCreateLabels(modifiedLabels sets.String) sets.String {
-	if len(modifiedLabels) == 0 {
-		return nil
-	}
-
-	forbiddenLabels := sets.NewString()
-	for label := range modifiedLabels {
-		namespace := getLabelNamespace(label)
-		// forbid kubelets from setting node-restriction labels
-		if namespace == v1.LabelNamespaceNodeRestriction || strings.HasSuffix(namespace, "."+v1.LabelNamespaceNodeRestriction) {
-			forbiddenLabels.Insert(label)
-		}
-	}
-	return forbiddenLabels
-}
-
-// getForbiddenLabels returns the set of labels that may not be set by the node on update.
-func (p *Plugin) getForbiddenUpdateLabels(modifiedLabels sets.String) sets.String {
+// getForbiddenLabels returns the set of labels that may not be added, removed, or modified by the node on create or update.
+func (p *Plugin) getForbiddenLabels(modifiedLabels sets.String) sets.String {
```

Comment:
With this change, kubelets can no longer set labels prefixed with `k8s.io` or `kubernetes.io`. Prior to kubernetes 1.19, this was considered acceptable behaviour, as `getForbiddenCreateLabels` did not set this restriction.

Is this an intentional change? If so, what is the recommended path forward for those using these prefixes? This appears to be a breaking change to the Node API.

For context, our team is working on [krustlet](https://github.com/deislabs/krustlet), and for WebAssembly modules to be correctly scheduled onto a krustlet node, we use node tolerations to ensure those pods are scheduled to the correct node architecture. These labels are set when the kubelet (krustlet) comes online and registers with the Node API.

https://github.com/deislabs/krustlet/blob/57604c635a3dab22fa4b0ac257868e7890634b85/demos/wasi/hello-world-c/k8s.yaml#L31-L34

Reply:
Yes, it was an intentional change, proposed [here](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/0000-20170814-bounding-self-labeling-kubelets.md#proposal) and rolled out from 1.13-1.19. The kubelet has not accepted the problematic labels since 1.16, so kube-apiserver started limiting labels on creation in 1.19, since the oldest supported kubelet that can register against a 1.19 API server is 1.17.

Changelogs:

https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.13.md#deprecations

> Use of the --node-labels flag to set labels under the kubernetes.io/ and k8s.io/ prefix will be subject to restriction by the NodeRestriction admission plugin in future releases. See [admission plugin documentation](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) for allowed labels.

https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.16.md#deprecations-and-removals

> kubelet removed the ability to set kubernetes.io- or k8s.io-prefixed labels via --node-labels, other than the [specifically allowed labels/prefixes.](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/0000-20170814-bounding-self-labeling-kubelets.md#proposal)

https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.19.md

> Kube-apiserver: The NodeRestriction admission plugin now restricts Node labels kubelets are permitted to set when creating a new Node to the --node-labels parameters accepted by kubelets in 1.16+.

Alternatives include:
* change to use `k8s.io/...` labels nodes are permitted to control - https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction
* change to use non-k8s.io labels
* disable the `NodeRestriction` admission plugin (not recommended, since this is what prevents nodes from modifying objects belonging to other nodes)